### DISCUSSION TITLE
resulted in a 400 Bad Request

### DISCUSSION THREAD

**Discussion by MostafaHamedAbdelmasoud** [imdhemy/laravel-in-app-purchases#85] [Q&A] [answered]
I am trying to verify receipt of purchase token and productID, but
it gives me an error resulted in  `400 Bad Request Invalid Valu (truncated...)

I have followed the steps in Readme and download the service account json file and attached it to my env file

but it gives me error 400 error, though it should work if I made everything in readme

**Comment by imdhemy** [comment]
Hi @MostafaHamedAbdelmasoud 
Could you please provide a code snippet?

A side question: why didn't you use the Bug template?

**Comment by imdhemy** [comment]
@MostafaHamedAbdelmasoud 
Excuse me Mostafa, but I meant a snippet from your code not a screenshot of the exception. This would help me to tell you how to troubleshoot the problem.

**Comment by MostafaHamedSpiderDAO** [comment]
no problem,

here is a snippet from my code

It's supposed to get subscription verification receipt from google play

![Screenshot from 2021-08-29 13-22-13](https://user-images.githubusercontent.com/88666247/131248831-1110cac4-0869-44ce-9c76-74e56cd58983.png)

**Comment by imdhemy** [comment, marked as answer]
@MostafaHamedSpiderDAO Okay, let's troubleshoot:

**N.B You can skip No. 1**

**1. Is the server authenticated?**
In your case, I think yes, because you are receiving `400` status not `401`. To double-check this, try to dump the value of your env variable `GOOGLE_APPLICATION_CREDENTIALS` using the `realpath()` function, if `false` is returned, you need to add the right path to the authentication JSON file.

**2. Is the item id valid?**
You can dump the value of `$itemId` and make sure that its value is one of the registered SKU values on your Google Play Console.

**3. Is the token valid?**
Double-check that the `$token` is a valid purchase token. Something like this one:
```
hfjblmhmkloppbpihhhifndn.AO-J1OxOKM878RHwS0Hl2Ti3WCvySGw9QTi5WtEUJHO4ppW7ai62vXtruAfOGFWFVdG8Spb3aJRnooesbP_Yfyo0_PXn6LCQ5g
```

**4. Is the android package name valid?**
The configured android package name should be the same as the one used during purchasing the subscription on your Android mobile application.

You can explicitly pass its value using:
```php
$productReceipt =  Subscription::googlePlay()
                       ->packageName('com.some.thing') // use the same value used on the Android APP
                       ->id($itemId)
                       ->token($token)
                       ->get();
```

**Finally**
- A successful response means a valid receipt.
- A Failed response means an invalid receipt is being verified, this explains why you may receive such exceptions.

**Comment by MostafaHamedAbdelmasoud** [comment]
thanks for your illustration, i will give it another try before we close this i have a one question:

item_id here is called subscription_monthly, could be this id worng here?

i mean should it be (org.spidervpn.app.subscription_monthly) or (subscription_monthly)?

**Comment by imdhemy** [comment]
@MostafaHamedAbdelmasoud Always happy to help!

No need to add the package name before the item id, so `subscription_monthly` is enough.

---

Good to mention the Google Application Credentials user should be the owner of the Android application. I'm not sure you are working on the spidervpn app or it was just an example. 

---

No need to close it, it's more a discussion than a bug. I'll transfer it to the discussion board.
